### PR TITLE
TFactor: Separate larft_gemv kernel and add a faster option

### PR DESCRIPTION
**Pull request by rasolca**
Note: currently still use the slow kernel.

Don't want to interfere with @albestro TFactor work.

### CODE REVIEW

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

File: miniapp/kernel/miniapp_larft_gemv.cpp [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2024, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#include <future>
```
Comment (msimberg) [inline miniapp/kernel/miniapp_larft_gemv.cpp:11]: Used?

File: miniapp/kernel/miniapp_larft_gemv.cpp [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+            elapsed_time_kernel = runner.runHandle(kernel_GPU0);
+            elapsed_time_post = runner.runStream(copy_tau_in_t);
+            break;
+          case 1:
+            elapsed_time_pre = runner.runStream(copy_tau);
+            elapsed_time_kernel = runner.runHandle(kernel_GPU1);
+            elapsed_time_post = runner.runStream(post_kernel_GPU1);
+            break;
+          default:
+            std::cout << "Error: Non existing kernel id" << opts.kernel_id << std::endl;
```
Comment (msimberg) [inline miniapp/kernel/miniapp_larft_gemv.cpp:160]:
```suggestion
            std::cout << "Error: Nonexistent kernel id" << opts.kernel_id << std::endl;
```
and then `std::terminate` or exit?

File: miniapp/kernel/miniapp_larft_gemv.cpp [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+                << " POST: " << elapsed_time_post << "s" << std::endl;
+
+      if ((opts.do_check == dlaf::miniapp::CheckIterFreq::Last && run_index == (opts.nruns - 1)) ||
+          opts.do_check == dlaf::miniapp::CheckIterFreq::All) {
+        auto t = createTile<T, Device::CPU>(el_t, {k, k}, k);
+        MC_reference(m, k, v.ptr(), v.ld(), tau.ptr(), t.ptr(), t.ld());
+
+        auto error = ts.check(t);
+        if (error > k)
+          std::cout << "CHECK FAILED!!!: ";
```
Comment (msimberg) [inline miniapp/kernel/miniapp_larft_gemv.cpp:192]: Is it this dramatic? :upside_down_face:
```suggestion
          std::cout << "CHECK FAILED!: ";
```
I see we are this dramatic in `miniapp_laset` as well, so either both change or neither changes... not a blocking comment.

File: test/unit/lapack/gpu/test_larft.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    {0, 0, 1, 1},       {0, 4, 4, 4},         {4, 0, 5, 5},                       // Empty
+    {6, 6, 6, 6},       {5, 3, 5, 3},         {7, 3, 10, 13},      {3, 7, 8, 8},  // Very Small
+    {127, 35, 128, 36}, {128, 128, 128, 128}, {96, 127, 256, 256},                // A bit larger
+};
+
+TYPED_TEST(LarftTestGPU, CorrectnessLocal) {
+  // Note:
+  // There are two main differences of GPU larft compared to LAPACK implementation:
+  // 1) the first element of the reflectors has to be set to 1. (implicit considered as one in LAPACK)
+  // 2) The lower part of the T factor is 0. (untouched in LAPACK)
```
Comment (msimberg) [inline test/unit/lapack/gpu/test_larft.cpp:48]: Is this an assumption that could be commented in `larft.cu` as well/instead?

**Review by msimberg — CHANGES_REQUESTED** [review]

File: miniapp/kernel/miniapp_larft_gemv.cpp (hunk 1 above)
Comment (rasolca) [reply miniapp/kernel/miniapp_larft_gemv.cpp:11]: Yes. see kernel_runner.h

**Comment by rasolca** [comment]
cscs-ci run

File: test/unit/lapack/gpu/test_larft.cpp (hunk 4 above)
Comment (rasolca) [reply test/unit/lapack/gpu/test_larft.cpp]: File sneaked in from #514... removed

File: miniapp/kernel/miniapp_larft_gemv.cpp (hunk 3 above)
Comment (rasolca) [reply miniapp/kernel/miniapp_larft_gemv.cpp:192]: As dramatic as https://github.com/eth-cscs/DLA-Future/blob/master/miniapp/kernel/miniapp_laset.cpp#L130

File: miniapp/kernel/miniapp_larft_gemv.cpp (hunk 1 above)
Comment (msimberg) [reply miniapp/kernel/miniapp_larft_gemv.cpp:11]: Why is it not included in `kernel_runner.h` then? It's not used directly in this file, or am I missing it somewhere?

Comment (rasolca) [reply miniapp/kernel/miniapp_larft_gemv.cpp]: I already moved there in the commit mentioned above.

File: miniapp/kernel/miniapp_larft_gemv.cpp (hunk 2 above)
Comment (rasolca) [reply miniapp/kernel/miniapp_larft_gemv.cpp]: https://github.com/eth-cscs/DLA-Future/pull/1219/files#diff-ddd8b6960bce12438abe598f5c25f1d5a2cddb794b5aeba723be222f9d717dddR160

**Comment by rasolca** [comment]
cscs-ci run

File: miniapp/kernel/miniapp_larft_gemv.cpp (hunk 1 above)
Comment (rasolca) [reply miniapp/kernel/miniapp_larft_gemv.cpp]: Ok, wrong commit... this is the correct one: 2caad6a1905387822bea49483c07f7eec78d6877

Comment (msimberg) [reply miniapp/kernel/miniapp_larft_gemv.cpp]: I see, makes more sense now. Thanks!

File: include/dlaf/gpu/blas/gpublas.h [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+
+#include <whip.hpp>
+
+#include <dlaf/gpu/blas/api.h>
+#include <dlaf/gpu/blas/error.h>
+#include <dlaf/util_cublas.h>
+
+#ifdef DLAF_WITH_HIP
+
+#include <pika/async_cuda/detail/cuda_event_callback.hpp>
```
Comment (msimberg) [inline include/dlaf/gpu/blas/gpublas.h:27]: This makes me sad, but it's our fault... At least there is a header included now. I think by now we have to accept that this is part of the public API. I've increased the priority of https://github.com/pika-org/pika/issues/887.

**Review by msimberg — APPROVED** [review]
Looks like ROCm pipeline is failing to build, but it's probably only a missing include or similar. This looks good to me otherwise.

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

**Comment by rasolca** [comment]
cscs-ci run

**Pull request merged**